### DISCUSSION TITLE
Node server: No exports found in module "./dist/main.js"

### DISCUSSION THREAD

**Discussion by ScreamZ** [vercel/vercel#4333] [General] [answered]
Hello,

Using exactly the same code it breaks with 502 error using nestJS deployment, it was working 6 days ago for my last deployment, based on that article https://trilon.io/blog/deploying-nestjs-to-zeit-now.

```
15:21:44:15
2020-05-11T13:21:51.030Z	8c5e1016-0d7c-4087-a2e6-e115551addeb	ERROR	No exports found in module "./dist/main.js".
2020-05-11T13:21:51.031Z	8c5e1016-0d7c-4087-a2e6-e115551addeb	ERROR	Did you forget to export a function or a server?
RequestId: 8c5e1016-0d7c-4087-a2e6-e115551addeb Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

It works with vercel dev…

Reproduction repository is here https://github.com/TrilonIO/zeit-now-nestjs

```
{
  "builds": [
    {
      "src": "dist/main.js",
      "use": "@vercel/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "dist/main.js"
    }
  ]
}
```

**Comment by styfle** [comment]
We just released [an update](https://github.com/zeit/now/releases/tag/vercel%4019.0.1) that should fix this issue.

I tested and it works correctly with your repo.

Also note that the Git Integration will fail to deploy because [dist/main.js](https://github.com/TrilonIO/zeit-now-nestjs/blob/master/dist/main.js) doesn't exist in your repository.

**Reply by ScreamZ** [reply]
Yes, I'm prebuilding on local, it's only used for staging environment for now, but production run in docker containers for the API part.

The frontend app is nextJS and run in vercel with github connected repository.

I can confirm it works with vercel@19.0.1.

Thanks 🙏🏻

**Comment by ScreamZ** [comment]
@styfle Did you changed something, looks like it breaks again…

Still the same error 502 with "no exported server" the code remains the same.

`vc --target staging --force`

**Comment by styfle** [comment]
No, nothing changed. Does it fail every time? What about if you leave out `--target`?

**Reply by ScreamZ** [reply]
@styfle Not better, you can have a look at my code repository, on vercel if you want to…

I made a support ticket about that:

Also here is my main.js

```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@nestjs/core");
const app_module_1 = require("./app.module");
async function bootstrap() {
    const app = await core_1.NestFactory.create(app_module_1.AppModule);
    await app.listen(3000);
}
bootstrap();
//# sourceMappingURL=main.js.map
```

EDIT: I tried, with the repo https://github.com/TrilonIO/zeit-now-nestjs it deploys correctly if I copy the dist folder from https://github.com/TrilonIO/zeit-now-nestjs into my project It also work well.

I'm using same tsconfig configuration on both (just tried to apply on https://github.com/TrilonIO/zeit-now-nestjs and worked), so it should be related to something else, but i have exactly the same main.js

**Reply by ScreamZ** [reply]
No issues with vc dev, everything goes well, but if I use the following config

```json
{
  "builds": [
    {
      "src": "src/main.ts",
      "use": "@vercel/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "src/main.ts"
    }
  ]
}
```

I have a crash with vc dev typescript compiler `Cannot find module 'src/auth/decorators/current-user.decorator'`

```ts
import { createParamDecorator, ExecutionContext } from "@nestjs/common";
import { GqlExecutionContext } from "@nestjs/graphql";

/**
 * Helper function for GQL to retrieve current user
 */
export const CurrentUser = createParamDecorator(
  (_data, context: ExecutionContext) => {
    return GqlExecutionContext.create(context).getContext().req.user;
  },
);
```

I'm importing it via `import { CurrentUser } from "src/auth/decorators/current-user.decorator";`

is it related to https://nextjs.org/blog/next-9-4#absolute-imports-and-aliases ? which has been just released ?

**Reply by ScreamZ** [reply]
Also my build log: 

```
15:57:25.209  Retrieving list of deployment files...
15:57:25.482  Downloading 108 deployment files...
15:57:34.084  Installing build runtime...
15:57:34.547  Build runtime installed: 462.648ms
15:57:35.382  Installing dependencies...
15:57:35.596  yarn install v1.22.0
15:57:35.666  [1/4] Resolving packages...
15:57:36.072  [2/4] Fetching packages...
15:57:49.376  info fsevents@2.1.3: The platform "linux" is incompatible with this module.
15:57:49.376  info "fsevents@2.1.3" is an optional dependency and failed compatibility check. Excluding it from installation.
15:57:49.377  info fsevents@1.2.12: The platform "linux" is incompatible with this module.
15:57:49.377  info "fsevents@1.2.12" is an optional dependency and failed compatibility check. Excluding it from installation.
15:57:49.378  [3/4] Linking dependencies...
15:57:49.382  warning "@nestjs/graphql > @nestjs/mapped-types@0.0.3" has unmet peer dependency "class-transformer@^0.2.3".
15:57:49.382  warning "@nestjs/graphql > @nestjs/mapped-types@0.0.3" has unmet peer dependency "class-validator@^0.11.1 || ^0.12.0".
15:57:49.383  warning "@nestjs/graphql > @apollo/federation > apollo-graphql@0.4.3" has incorrect peer dependency "graphql@^14.2.1".
15:57:49.386  warning "@nestjs/graphql > @apollo/gateway > apollo-server-core > graphql-tag@2.10.3" has incorrect peer dependency "graphql@^0.9.0 || ^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.0 || ^14.0.0".
15:57:49.386  warning "@nestjs/graphql > @apollo/gateway > apollo-server-core > graphql-upload@8.1.0" has incorrect peer dependency "graphql@0.13.1 - 14".
15:57:49.386  warning "apollo-server-express > subscriptions-transport-ws@0.9.16" has incorrect peer dependency "graphql@^0.10.0 || ^0.11.0 || ^0.12.0 || ^0.13.1 || ^14.0.2".
15:57:49.387  warning "@nestjs/graphql > @apollo/gateway > apollo-server-core > graphql-tools > apollo-utilities@1.3.3" has incorrect peer dependency "graphql@^0.11.0 || ^0.12.0 || ^0.13.0 || ^14.0.0".
15:57:49.388  warning "apollo-server-express > graphql-subscriptions@1.1.0" has incorrect peer dependency "graphql@^0.10.5 || ^0.11.3 || ^0.12.0 || ^0.13.0 || ^14.0.0".
15:57:57.840  [4/4] Building fresh packages...
15:57:59.003  Done in 23.41s.
15:58:13.337  Uploading build outputs...
15:58:16.352  Build completed. Populating build cache...
15:58:30.216  Uploading build cache [34.20 MB]...
15:58:30.896  Build cache uploaded: 679.434ms
15:58:30.909  Done with "dist/main.js"
```

**Reply by styfle** [reply]
You're not using `@vercel/next`, you're using `@vercel/node` so Next.js changes are unrelated.

If copying the dist folder from one to the other works, it sounds like your build step is outputting different files and is likely the source of the bug.

**Reply by ScreamZ** [reply]
Also tried with such .vercelignore

```
src
tsconfig.json
tsconfig.build.json
```

Not working

**Reply by ScreamZ** [reply]
@styfle Alright but how can I track errors, as nothing shows in now console instead of 

```
15:21:44:15
2020-05-11T13:21:51.030Z	8c5e1016-0d7c-4087-a2e6-e115551addeb	ERROR	No exports found in module "./dist/main.js".
2020-05-11T13:21:51.031Z	8c5e1016-0d7c-4087-a2e6-e115551addeb	ERROR	Did you forget to export a function or a server?
RequestId: 8c5e1016-0d7c-4087-a2e6-e115551addeb Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

And no issue on vercel-dev… with js at least

**Reply by ScreamZ** [reply]
I've reached you on twitter so we can exchange smoothly if you want, thanks for helping 🙏🏻

**Comment by ScreamZ** [comment]
Is there any restriction in terms of file size or build duration that could fail? I don't have enough tools to investigate on my own why this is crashing on now and not on local with vc dev @styfle  :(

**Reply by styfle** [reply]
There are limits but these would fail before earlier. https://vercel.com/docs/v2/platform/limits

Your deployment is failing at runtime so its not related to limits. It's related to the missing export in `./dist/main.js`.

**Comment by styfle** [comment, marked as answer]
We don't support async `server.listen()`. If you're using the ["serverful"](https://vercel.com/docs/runtimes#advanced-usage/advanced-node-js-usage/legacy-serverful-behavior) API then it needs to be listening when the main function is done executing. Alternatively, you can export the server `module.exports = server` and let Vercel listen.

I suspect the reason you see that it works sometimes and that it doesn't work other times depends on how quickly your async function initializes and starts listening.

Take a look at this Koa example to see a solution that will always work: https://github.com/zeit/now/blob/master/packages/now-node/test/fixtures/16-servers/koa/index.js

**Reply by ScreamZ** [reply]
Yes, you're right, it should be about that. I guess that when the app is getting heavier and heavier and initialization is taking more time now, while initialization is async it now fails.

Using Nest I need to await for `NestFactory.create(AppModule);` in order to resolve dependency injections at start.

So about 

```
  app.listen(process.env.PORT || 3000);
  app.listenAsync(process.env.PORT || 3000);
```
We don't care that much.

I guess I'll give up on that and switch on some Google App Engine for such backend or Heroku, it will be easier. I'll keep my frontend on now, with my nextJS app.

They made an awesome https://github.com/nestjs/azure-func-http lib for wiring azure function even with the azure router in case you don't rely on express, I'll check if I can build something similar for now in case of prototyping apps, but in production, this is clearly not designed for cloud functions. :)

Thanks @styfle

**Comment by aecorredor** [comment]
@styfle I'm having the same issue for a express server that needs to get two databases URLs from heroku (they recommend not hardcoding these since they can change without notice). The error happens very randomly so it's probably due to what was discussed above. The code looks like this:

```js
Promise.all([
  heroku.getConfigVar('DATABASE_URL'),
  heroku.getConfigVar('REDISCLOUD_URL'),
]).then(([dbUrl, redisUrl]) => {
  // Both of these env variables need to be set before importing the setup file
  process.env.DATABASE_URL = dbUrl;
  process.env.REDIS_URL = redisUrl;

  const setUpApp = require('./setup').default;
  const app = express();
  const models = setUpApp(app) as Models;

  // ...some more code not relevant to the example 

  const server = http.createServer(app);

  models.sequelize.sync().then(() => {
    server.listen({ port }, () => {
      console.log(`🚀 Server is ready at ${protocol}://${origin}:${port}`);
    });
  });
});
```

I'm also running `sequelize.sync()`, but I can remove that since it's not really suited for production.

I was looking at https://github.com/zeit/now/blob/ba25004ea8e69c9adf17b696376a1df9fe97166b/packages/now-node/src/launcher.ts#L67 and see that the timeout is 5 seconds. Maybe increasing that a little more could be convenient for this type of use case? or is there a reason that it has to be 5 seconds?

If not, I guess I can just hardcode the database URLs as vercel secrets for now and re-deploy if heroku decides to change URLs for some reason.

**Reply by ScreamZ** [reply]
As discussed above, I guess it's not scalable to play with monolithic apps on now (vercel), you should (as I did) switch to Heroku for free tier apps, any container-based or server, like Google App engine.

Otherwise at some point you'll get stuck as I was

**Reply by styfle** [reply]
Consider adding the environment variables to Vercel which will be available to the function without any additional setup via `process.env` https://vercel.com/blog/environment-variables-ui

**Reply by aecorredor** [reply]
@styfle the problem with using environment variables for heroku database URLs is that they can change at any point on heroku’s side, and that’s why they advise to obtain them at runtime using their API. If we were to store them on vercel’s environment variables and have to change them at some point,  we would have to redeploy the service correct?

**Reply by styfle** [reply]
Correct, environment variables are only applied at deploy time and are burned into the deployment.

It sounds like Heroku's database is not a good fit for serverless functions.

We also discourage the use of Express which is not a good fit for serverless functions.

Instead, you can use [vercel helpers](https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-request-and-response-objects) that are available without additional dependencies.

**Reply by aecorredor** [reply]
Man, I miss your docker offering :(
